Document how to connect ADXL345 to Pico then on to Pi.

yegct opened the pull request:
This is a documentation-only change, no code is modified.

This pull request describes how to hook up an ADXL345 accelerometer to a Raspberry Pi Pico and then on to the main Raspberry Pi.

The documentation already describes [how to wire up an MPU-6000/9000 with Pico](https://www.klipper3d.org/Measuring_Resonances.html#configure-mpu-60009000-series-with-pico), so I figured it would be useful to describe this with the ADXL345. I expect this is the more commonly used accelerometer. Additionally, it's helpful to have your accelerometer connect to the main Raspberry Pi via USB instead of GPIO as it's a lot more convenient to move between klipper setups.

This is based on documentation from [this discourse post](https://klipper.discourse.group/t/raspberry-pi-pico-adxl345-portable-resonance-measurement/1757), with my thanks.

Signed-off-by: Christopher Thompson <chris@hypocrite.org>

KevinOConnor commented:
Thanks.  I'd like to see if there are comments from @dingyifei , @dmbutyugin , @Sineos on this change.

-Kevin

thijstriemstra reviewed (approved):
Had to follow some youtube clips (nero3d) to get this working yesterday, would've been nice to see documentation like this. Looks good to me!

The image for `make menuconfig` could be updated for latest klipper though, it now also includes a bootloader option (that does not have to be changed though).

yegct commented:
> The image for `make menuconfig` could be updated for latest klipper though, it now also includes a bootloader option (that does not have to be changed though) that is not present in this screenshot.

Very happy to update the screenshot. I should be able to do so tomorrow! Thanks for the review.

Sineos commented on docs/Measuring_Resonances.md:
```diff
@@ -60,6 +58,32 @@ Fritzing wiring diagrams for some of the ADXL345 boards:
+accelerometer on other Klipper devices, as you can connect via USB instead
+of GPIO.
+
+Please be particularly careful with your ADXL345. You want one without a
+level shifter and without being forced into I2C mode. For more details,
+see [this discourse post](https://klipper.discourse.group/t/raspberry-pi-pico-adxl345-portable-resonance-measurement/1757).
```
I think it should be reworded, as it is OK to use an ADXL with level shifter as long as you do not connect it to VCC 5V.

Maybe something like:

> In order to avoid damage to your RPi make sure to connect the ADXL345 to 3.3V only. Depending on the board's layout, a level shifter may be present, which makes 5V dangerous for your RPi.

Sineos commented on docs/Measuring_Resonances.md:
```diff
@@ -60,6 +58,32 @@ Fritzing wiring diagrams for some of the ADXL345 boards:
+level shifter and without being forced into I2C mode. For more details,
+see [this discourse post](https://klipper.discourse.group/t/raspberry-pi-pico-adxl345-portable-resonance-measurement/1757).
+
+Note that we do not hook up the INT1 or INT2 pins on the ADXL345.
```
Needed?

thijstriemstra replied:
it doesn't hurt or confuse me as a user to have it in there.

Sineos commented on docs/Measuring_Resonances.md:
````diff
@@ -164,6 +188,66 @@ probe_points:
 
+#### Configure ADXL345 With Pi Pico
+
+##### Flash the Pico Firmware
+
+On your Raspberry Pi, compile the firmware for the Pico.
+
+```
+cd ~/klipper
+make clean
+make menuconfig
+```
+![Pico menuconfig](img/klipper_pico_menuconfig.png)
+```
+make
+```
+
+Now, while holding down the `BOOTSEL` button on the Pico, connect the Pico to
+the Raspberry Pi via USB. A new block device should appear. If you have no
+other block devices, this will be `/dev/sda`. You should now copy the klipper
+firmware to the pico.
+
+```
+sudo mount /dev/sda1 /mnt
+sudo cp out/klipper.uf2 /mnt
+sudo umount /mnt
+```
````
Meanwhile outdated. See https://github.com/Klipper3d/klipper/commit/7c0559c6e62506af73d0e8f22733615705664dd5

Sineos reviewed (changes requested):
The connection diagram is quite hard to read. I think a bit higher resolution would be better. See https://canada1.discourse-cdn.com/free1/uploads/klipper/original/1X/9c8eaa478031620380eda99f012b4046ddc96e2f.png (this is the original from discourse)

yegct commented:
Thanks for the feedback, much appreciated! Converting this PR to draft until I get a chance to make the updates.

yegct commented:
> The connection diagram is quite hard to read. I think a bit higher resolution would be better. See https://canada1.discourse-cdn.com/free1/uploads/klipper/original/1X/9c8eaa478031620380eda99f012b4046ddc96e2f.png (this is the original from discourse)

When I push my changes, it will have the higher resolution image.

yegct commented:
@Sineos , I believe I have addressed all of your concerns in your first review pass of this PR. Back to you!

dingyifei commented:
I think the section `Flash the Pico Firmware` document fits better under the `Device Specific Documents`, just like the `RPi microcontroller` document. It is also worth mentioning the SPI performance may degrade if the rp2040 is on a controller board in charge of more than the accelerometer. I remember someone posting their logic analyzer result showing fewer reads and different frequency results when they used an STM32 F4 on a controller board.

Sineos commented on docs/Measuring_Resonances.md:
````diff
@@ -164,6 +187,62 @@ probe_points:
+```
+![Pico menuconfig](img/klipper_pico_menuconfig.png)
+
+Now, while holding down the `BOOTSEL` button on the Pico, connect the Pico to
+the Raspberry Pi via USB. Compile and flash the firmware.
````
I'm not sure, maybe @KevinOConnor can comment but I believe a 
`sudo apt-install pkg-config`
is needed

Sineos replied:
Fine for me. I always use vanilla Linux distributions and I had to install it manually.

Sineos commented on docs/Measuring_Resonances.md:
```diff
@@ -228,7 +307,9 @@ Recv: // adxl345 values (x, y, z): 470.719200, 941.438400, 9728.196800
 
 If you get an error like `Invalid adxl345 id (got xx vs e5)`, where `xx`
-is some other ID, it is indicative of the connection problem with ADXL345,
+is some other ID, immediately try again. There's an issue with SPI
```
AFAIK the first measurement with a Pico will always fail. Maybe it should be in the section of the Pico itself

joaobarros replied:
Indeed but it wasn't always the case:  https://github.com/Klipper3d/klipper/issues/4712

yegct replied:
I believe it's a bug as pointed out by joaobarros. I don't believe it's Pico-specific. For that reason, I don't believe it should be in the Pico section. Strong opinion, loosely held.

Sineos replied:
Well, if a bug or not: I only see this behavior with the Pico. Not on a RPi 3 and also not on a Orange Pi 3 LTS

yegct replied:
Fascinating. After this PR, I may take a stab at fixing the bug. I don't see how it would be Pico-specific or Pi 4 specific but maybe I'll find out.

glabifrons replied:
It's specifically a bug in adxl345.py. A function to initialize the chip was removed in commit https://github.com/Klipper3d/klipper/commit/8cfaed441d1ec295c8c4be8e0733099fe3f3395a
I don't pretend to understand the reasoning behind the change, but that is the change that introduced the "always fails the first time" bug.

yegct replied:
Thanks, glibifrons. That at least should make it easy to reproduce and examine, whether that's done by me or by someone else. :)

thijstriemstra replied:
@glabifrons could you open a ticket for this? It's an annoying bug that should be fixed if possible, especially if it's simply reverting part of a commit.

glabifrons replied:
@thijstriemstra I did, over a year ago: https://github.com/Klipper3d/klipper/issues/4712

Sineos reviewed (commented):
Thanks for working on it and sorry for the late reply.

Regarding @dingyifei comment:
Both locations have their pros and cons. I have no feelings towards either.

yegct commented:
Thanks for the feedback. I'll make the updates when I have a moment, sometime in the next few days.

thijstriemstra commented on docs/Measuring_Resonances.md:
```diff
@@ -60,6 +58,31 @@ Fritzing wiring diagrams for some of the ADXL345 boards:
 
 ![ADXL345-Rpi](img/adxl345-fritzing.png)
 
+###### Via Raspberry Pi Pico
```
Instead of `Via` -> `Using` (or leave it out all together)

yegct commented:
> Thanks for working on it and sorry for the late reply.
> 
> Regarding @dingyifei comment: Both locations have their pros and cons. I have no feelings towards either.

@dingyifei, I added a bit about only running the accelerometer. After careful consideration, I believe the content belongs here rather than in a device-specific document. This is because this document already contains similar information (e.g. other accelerometers such as the MPU-9250, e.g. installing software packages, e.g. hooking up an accelerometer directly to a raspberry pi, instead of via a pi pico).

yegct commented:
@thijstriemstra, @Sineos , others: I believe I have made all requested changes or justified why the changes should not be made. Back to you folks for another pass.

(Also brought back up to date with the main branch)

Sineos commented on docs/Measuring_Resonances.md:
````diff
@@ -164,6 +191,62 @@ probe_points:
+now add an `adxl.cfg` file with the following settings:
+
+```
+[mcu adxl]
+# Change <mySerial> to whatever you found above. For example,
+# usb-Klipper_rp2040_E661640843545B2E-if00
+serial: /dev/serial/by-id/usb-Klipper_rp2040_<mySerial>
+
+[adxl345]
+cs_pin: adxl:gpio1
+spi_bus: spi0a
+axes_map: x,z,y
+
+[resonance_tester]
+accel_chip: adxl345
+probe_points:
+    # Somewhere slightly above the middle of your print bed
+    147,154, 20
````
There is one thing I forgot to mention (also on the thread in discourse)
Tests have shown that the 3V3 line on the Pico is very noisy, which leads to higher noise values on the ADXL measurements as well. This behavior can be improved by setting

```
[output_pin power_mode]
pin: adxl:gpio23
```

This will force the Pico's power source to go into high-performance mode rather than power saving mode.

yegct replied:
This (more or less) matches the configuration described [elsewhere in the document](https://www.klipper3d.org/Measuring_Resonances.html#configure-mpu-60009000-series-with-pico). I will go and test to see if the above works.

Sineos replied:
True, it was added by an excellent documentation rework of @dingyifei but it applies to this as well

yegct replied:
Tested and change pushed. Seems to at least do no harm and, as you point out, may meaningfully improve things.

Sineos reviewed (approved):
LGTM 👍 

If you want you can work in the comment about the Pico's power mode. It makes a difference when looking at the raw data. If it makes a difference for real life measurements I'm not sure.

KevinOConnor commented:
Thanks.  In general it looks fine to me.  However, I'm unsure about adding a link to the Discourse forums in the main documentation.  (Over time, forum posts tend to "ramble" as new people comment, and I think it may be confusing to a new user if they are directed to a potentially rambling forum post.)

-Kevin

yegct commented:
> Thanks. In general it looks fine to me. However, I'm unsure about adding a link to the Discourse forums in the main documentation. (Over time, forum posts tend to "ramble" as new people comment, and I think it may be confusing to a new user if they are directed to a potentially rambling forum post.)

@KevinOConnor, reasonable. I have removed the link and brought the branch back up to date with the `master` branch. If you prefer the entire paragraph be removed, let me know. It wasn't a concern for me, my ADXL345 was correct. I'm unsure if my experience would generalise, though.

thijstriemstra reviewed (approved):
thanks! and thanks for sticking to it, addressing review comments.

yegct commented:
> 

@KevinOConnor , this got closed as inactive. I believe I've addressed all of the concerns and the ticket has two signoffs. Are you waiting for anything more from me?

KevinOConnor commented:
Sorry, it slipped on my side.  I committed this change now.

Thanks.
-Kevin
